File: leetcode/0001-1000/0101-0200/0198-house-robber/HouseRobber.java
public class HouseRobber {
  public static void main(String[] args) {
    System.out.println(new Solution().rob(new int[] { 2, 7, 9, 6, 4 }));
  }
}

class Solution {
  public int rob(int[] numbers) {
    int previous = 0;
    int current = 0;

    for (int number : numbers) {
      int a = Math.max(previous, current);
      int b = previous + number;

      //
      previous = a;
      current = b;
    }

    return Math.max(previous, current);
  }
}

File: leetcode/0001-1000/0101-0200/0191-0200/0198-house-robber/HouseRobber.ts
function rob(nums: number[]): number {
  let [previous, current] = [0, 0];

  for (const num of nums) {
    [previous, current] = [Math.max(previous, current), previous + num];
  }

  return Math.max(previous, current);
};

console.log(rob([2, 7, 9, 6, 1]));

// 1 2 3 4 5 6 7 8 9
//
// x   x   x   x   x
// x   x   x     x  
// x   x     x   x  
// x   x     x     x
//
//
// x     x   x   x
// x     x   x     x
// x     x     x   x
